### PR TITLE
Machine translation model based on RNN

### LINKED ISSUE #674: Need provide a demo for actually usable beam search
- [x] Validate the weights sharing between training and inference
- [ ] Add complex RNN model (with attention)
- [x] Enhance the sequence_expand operator

### PR DESCRIPTION
**Pull request by ktlichkid**
Resolved #674

### CODE REVIEW

File: fluid/neural_machine_translation/rnn_search/attention_model.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    "--batch_size",
+    type=int,
+    default=32,
+    help="The sequence number of a mini-batch data. (default: %(default)d)")
+parser.add_argument(
+    "--dict_size",
+    type=int,
+    default=30000,
+    help="The dictionary capacity. Dictionaries of source sequence and "
+    "target dictionary have same capacity. (default: %(default)d)")
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/attention_model.py]: `source sequence` -> `source language`
`target dictionary have same capacity` -> `target language have the same capacity`

File: fluid/neural_machine_translation/rnn_search/attention_model.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    type=float,
+    default=0.01,
+    help="Learning rate used to train the model. (default: %(default)f)")
+parser.add_argument(
+    "--infer_only", action='store_true', help="If set, run forward only.")
+parser.add_argument(
+    "--beam_size",
+    type=int,
+    default=3,
+    help="The width for beam searching. (default: %(default)d)")
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/attention_model.py]: `beam searching` -> `beam search`

File: fluid/neural_machine_translation/rnn_search/attention_model.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+parser.add_argument(
+    "--use_gpu",
+    type=distutils.util.strtobool,
+    default=True,
+    help="Whether to use gpu. (default: %(default)d)")
+parser.add_argument(
+    "--max_length",
+    type=int,
+    default=50,
+    help="The maximum length of sequence when doing generation. "
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/attention_model.py]: ` when doing` -> `in generation`

File: fluid/neural_machine_translation/rnn_search/attention_model.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+    place = core.CUDAPlace(0) if args.use_gpu else core.CPUPlace()
+    exe = Executor(place)
+    exe.run(framework.default_startup_program())
+
+    feed_list = [
+        main_program.global_block().var(var_name) for var_name in feed_order
+    ]
+    feeder = fluid.DataFeeder(feed_list, place)
+
+    def do_validation():
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/attention_model.py]: `do_validation` -> `validation`

File: fluid/neural_machine_translation/rnn_search/no_attention_model.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+max_length = args.max_length
+pass_num = args.pass_num
+save_interval = args.save_interval
+IS_SPARSE = True
+topk_size = 50
+beam_size = args.beam_size
+learning_rate = args.learning_rate
+infer_only = args.infer_only
+place = core.CUDAPlace(0) if args.use_gpu else core.CPUPlace()
+model_save_dir = args.save_dir
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/no_attention_model.py]: Just use `args`, and simplify these lines

File: fluid/neural_machine_translation/rnn_search/no_attention_model.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    ]
+    feeder = fluid.DataFeeder(feed_list, place)
+
+    for pass_id in range(1, pass_num + 1):
+        for batch_id, data in enumerate(train_reader()):
+            outs = exe.run(program,
+                           feed=feeder.feed(data),
+                           fetch_list=[avg_cost])
+            avg_cost_val = np.array(outs[0])
+            if pass_id % 1 == 0:
```
Comment (kuke) [inline fluid/neural_machine_translation/rnn_search/no_attention_model.py]: Can remove #L251

**Review by kuke — COMMENTED** [review]
Please separate model config, train and infer for code reuse.

File: fluid/neural_machine_translation/rnn_search/attention_model.py (hunk 1 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/attention_model.py]: Done

File: fluid/neural_machine_translation/rnn_search/attention_model.py (hunk 2 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/attention_model.py]: Done

File: fluid/neural_machine_translation/rnn_search/attention_model.py (hunk 3 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/attention_model.py]: Done

File: fluid/neural_machine_translation/rnn_search/attention_model.py (hunk 4 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/attention_model.py]: Done

File: fluid/neural_machine_translation/rnn_search/no_attention_model.py (hunk 5 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/no_attention_model.py]: Done

File: fluid/neural_machine_translation/rnn_search/no_attention_model.py (hunk 6 above)
Comment (ktlichkid) [reply fluid/neural_machine_translation/rnn_search/no_attention_model.py]: Done

**Review by kuke — APPROVED** [review]
LGTM

**Pull request merged**